codec-http2/src/main/java/io/netty/handler/codec/http2/ClearTextHttp2ServerUpgradeHandler.java

```diff
@@ -0,0 +1,88 @@
+import io.netty.handler.codec.ByteToMessageDecoder;
+import io.netty.handler.codec.http.HttpServerCodec;
+import io.netty.handler.codec.http.HttpServerUpgradeHandler;
+import io.netty.handler.codec.http.HttpServerUpgradeHandler.UpgradeCodecFactory;
+
+import java.util.List;
+
+import static io.netty.buffer.Unpooled.unreleasableBuffer;
+import static io.netty.handler.codec.http2.Http2CodecUtil.connectionPrefaceBuf;
+
+/**
+ * Providing how starting http2 protocol, by h2c HTTP upgrade, or by h2 with Prior Knowledge.
+ * This handler will set up {@link PriorKnowlegeHandler}, {@link HttpServerCodec},
+ * and {@link UpgradeCodecFactory} into the pipeline when handler added.
+ * And will set up pipeline correctly once it detect the protocol is h2c or h2.
+ */
+public class ClearTextHttp2ServerUpgradeHandler extends ChannelHandlerAdapter {
+    private static final ByteBuf CONNECTION_PREFACE = unreleasableBuffer(connectionPrefaceBuf());
+
+    private final HttpServerCodec httpServerCodec;
+    private final ChannelHandler priorKnowledgeUpgradeHandler;
+    private final HttpServerUpgradeHandler httpServerUpgradeHandler;
+
+    public ClearTextHttp2ServerUpgradeHandler(HttpServerCodec httpServerCodec,
+                                              HttpServerUpgradeHandler httpServerUpgradeHandler,
+                                              ChannelHandler priorKnowledgeUpgradeHandler) {
+        this.httpServerCodec = httpServerCodec;
+        this.httpServerUpgradeHandler = httpServerUpgradeHandler;
+        this.priorKnowledgeUpgradeHandler = priorKnowledgeUpgradeHandler;
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+        ctx.pipeline()
+           .addBefore(ctx.name(), null, new PriorKnowlegeHandler())
+           .addBefore(ctx.name(), null, httpServerCodec)
+           .addBefore(ctx.name(), null, httpServerUpgradeHandler)
+           .remove(this);
+    }
+
+    class PriorKnowlegeHandler extends ByteToMessageDecoder {
+        @Override
+        protected void decode(ChannelHandlerContext ctx, ByteBuf in, List<Object> out) throws Exception {
+            int prefaceLength = CONNECTION_PREFACE.readableBytes();
+            int bytesRead = Math.min(in.readableBytes(), prefaceLength);
+
+            if (!ByteBufUtil.equals(CONNECTION_PREFACE, CONNECTION_PREFACE.readerIndex(),
+                                    in, in.readerIndex(), bytesRead)) {
+                out.add(in.readBytes(in.readableBytes()));
+                ctx.pipeline().remove(this);
+            } else if (bytesRead == prefaceLength) {
+                // Full h2 preface match, removed source codec, using http2 codec to handle
+                // following network traffic
+                ctx.pipeline().remove(httpServerCodec);
+                ctx.pipeline().remove(httpServerUpgradeHandler);
+
+                ctx.pipeline().addAfter(ctx.name(), null, priorKnowledgeUpgradeHandler);
```

Comment:
it seems like this is installed with the name `null` by `ClearTextHttp2ServerUpgradeHandler`.  how does this work?